### PR TITLE
(:sparkles:)Scaffold conversion webhook template (v3-alpha plugin)

### PR DESCRIPTION
**Pull request by varshaprasad96**
This PR adds scaffolding when conversion webhook is specified in the
`kubebuilder create webhook` command of kubebuilder.

Example usage:
`kubebuilder create webhook  --group crew --version v1 --kind FirstKind --conversion --hub v1 --spoke v2`

1. FirstKind_conversion.go scaffolded inside `api/v1/firstmate_webhook.go` having [`conversion.Hub`](https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/conversion/conversion.go#L37) interface.
2. FirstKind_conversion.go scaffolded inside `api/v2/firstmate_webhook.go` having method signatures to implement [`conversion.convertible`](https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/conversion/conversion.go#L27) interface.
3. Webhook registered in `main.go`

Related to issue: #1663 

c/c: @estroz

### CODE REVIEW

File: pkg/plugin/v3/webhook.go
Comment (camilamacedo86) [inline pkg/plugin/v3/webhook.go:111]: Why should we make mandatory inform them?

File: go.mod
Comment (camilamacedo86) [inline go.mod:9]: It shows not be part of the scope of this pr. Could you please revert?

**Review by camilamacedo86 — CHANGES_REQUESTED** [review]
Hi @varshaprasad96, 

Really nice contribution 🥇  just a few points. Please, check the comments:
- It is not passing in the check testdata because you add the args as mandatory. See that you will need change the script that generates the samples in the testdata: https://github.com/kubernetes-sigs/kubebuilder/blob/master/generate_testdata.sh 
- It is missing be covered by tests. 
- Also, we need to update its docs

File: pkg/plugin/v3/webhook.go
Comment (varshaprasad96) [reply pkg/plugin/v3/webhook.go:111]: Because it depends on the user on what version they want to have as the hub. Without it, I think it wouldn't be fair to randomly chose a hub version and a non-storage version.

Comment (camilamacedo86) [reply pkg/plugin/v3/webhook.go:111]: Can we keep the current behaviour when these values are not informed?

File: go.mod
Comment (camilamacedo86) [inline go.mod:15]: could you please remove the indirect ones?

File: pkg/plugin/v3/webhook.go
Comment (camilamacedo86) [inline pkg/plugin/v3/webhook.go:108]: Is it a warn/info? I do not think that it is required,

File: pkg/plugin/v3/webhook.go
Comment (camilamacedo86) [inline pkg/plugin/v3/webhook.go:115]: I think we need to start to cover the code with tests. WDYT about add a webhook_test here to check these scenarios?

File: pkg/plugin/v3/scaffolds/internal/templates/config/api/spoke_conversion.go
Comment (camilamacedo86) [inline pkg/plugin/v3/scaffolds/internal/templates/config/api/spoke_conversion.go:63]:
```suggestion
	// Implement your logic here to convert from hub to spoke version.
```

Comment (camilamacedo86) [reply pkg/plugin/v3/scaffolds/internal/templates/config/api/spoke_conversion.go:63]: To be similar to: https://github.com/kubernetes-sigs/kubebuilder/blob/master/testdata/project-v3/controllers/admiral_controller.go#L44

File: pkg/plugin/v3/scaffolds/internal/templates/config/api/spoke_conversion.go
Comment (camilamacedo86) [inline pkg/plugin/v3/scaffolds/internal/templates/config/api/spoke_conversion.go:69]:
```suggestion
	// Implement your logic here to convert from spoke to hub version.
```

File: pkg/plugin/v3/scaffolds/internal/templates/config/api/spoke_conversion.go
Comment (camilamacedo86) [inline pkg/plugin/v3/scaffolds/internal/templates/config/api/spoke_conversion.go:60]: hav you check this in any file scaffolded? is this an standard in the project?

**Review by camilamacedo86 — CHANGES_REQUESTED** [review]
Really great contribution. Following a few things that needs to be done here:
- [ ] the title of the pr requires the mojo. See: https://github.com/kubernetes-sigs/kubebuilder/blob/master/CONTRIBUTING.md#pr-process
- [ ] revert the changes in the go mod that shows not be part of its scope. 
- [ ] the file generate_testdata.sh needs to be a supplement to gen in the tests data this scenario as well for the v3-projects. 
- [ ] we need to update the docs to describe the new options and how it works. 
- [ ] could we add this scenario to the e2e tests? 
- [ ] some nits in the code
- [ ] rebase with master

File: pkg/plugin/v3/scaffolds/internal/templates/config/api/hub_conversion.go
Comment (camilamacedo86) [inline pkg/plugin/v3/scaffolds/internal/templates/config/api/hub_conversion.go:47]: IHMO we should not add prints here. Have we any other boilerplate printing. the actions?
If we would like to log out the action then, I think it needs to be in the plugin which is reasonable for the execution. 
The boilerplate just define the template which will be executed. so, I do not think that it is part of its domain of responsibility.

File: pkg/plugin/v3/scaffolds/internal/templates/config/api/hub_conversion.go
Comment (camilamacedo86) [inline pkg/plugin/v3/scaffolds/internal/templates/config/api/hub_conversion.go:63]: Should we scaffold the project with or it should be in the docs?

File: pkg/plugin/v3/webhook.go
Comment (varshaprasad96) [reply pkg/plugin/v3/webhook.go:108]: This is an additional info informing users that the feature is available for them to set.

**Comment by prafull01** [comment]
Once this PR gets merged and  conversion tests are supported in envtest ([PR here](https://github.com/kubernetes-sigs/kubebuilder/pull/1710/files#r500513284)), the test suite scaffolding is also needs to be done. For More details: [https://github.com/kubernetes-sigs/kubebuilder/pull/1710/files#r500513284](url)

File: pkg/plugin/v3/scaffolds/internal/templates/config/api/spoke_conversion.go
Comment (varshaprasad96) [reply pkg/plugin/v3/scaffolds/internal/templates/config/api/spoke_conversion.go:60]: This appears we scaffold admission/mutating WebHooks too. So I had added it here.

File: pkg/plugin/v3/scaffolds/internal/templates/config/api/hub_conversion.go
Comment (varshaprasad96) [reply pkg/plugin/v3/scaffolds/internal/templates/config/api/hub_conversion.go:47]: Had added this, because webhook conversion scaffolding prints the path: [here](https://github.com/kubernetes-sigs/kubebuilder/blob/49f7a5eb70f85034473ff18c85827337f3c7d9a9/pkg/plugin/v3/scaffolds/internal/templates/config/api/webhook.go#L59). Shall I remove it or still have it ? Im fine with removing it too.

File: pkg/plugin/v3/scaffolds/internal/templates/config/api/hub_conversion.go
Comment (varshaprasad96) [reply pkg/plugin/v3/scaffolds/internal/templates/config/api/hub_conversion.go:63]: This is as present in the example conversion WH operator - https://github.com/kubernetes-sigs/kubebuilder/blob/master/docs/book/src/multiversion-tutorial/testdata/project/api/v1/cronjob_conversion.go#L19-L24. Do we have it or shall I remove it?

File: pkg/plugin/v3/webhook.go
Comment (camilamacedo86) [reply pkg/plugin/v3/webhook.go:108]: IMO we ough to not use the `fmt.Println` to print WARN/INFO logs for the tool. We can use either logr with zapr, or klog v2

File: pkg/plugin/v3/scaffolds/internal/templates/config/api/hub_conversion.go
Comment (estroz) [inline pkg/plugin/v3/scaffolds/internal/templates/config/api/hub_conversion.go:66]: This doesn't necessarily need to be a pointer since this method does nothing.
```suggestion
func ({{ .Resource.Kind }}) Hub() {}
```

File: pkg/plugin/v3/webhook.go
Comment (estroz) [inline pkg/plugin/v3/webhook.go:107]: This condition encompasses the above one logically, can you combine them?

File: pkg/plugin/v3/webhook.go
Comment (estroz) [inline pkg/plugin/v3/webhook.go:82]: Can we simplify these flags by removing `--conversion`, and treating setting `conversion: true` as `--hub --spoke`, seeing as:
- `--hub --spoke` always requires `--conversion` always be set
- there's no point in setting `--conversion` without `--hub --spoke`

File: pkg/plugin/v3/scaffolds/webhook.go
Comment (estroz) [inline pkg/plugin/v3/scaffolds/webhook.go:52]:
```suggestion
	webhookSpokeVersion string,
```
and ditto below.

**Review by estroz — CHANGES_REQUESTED** [review]
Mostly nits, one major suggestion regarding flag consolidation.

**Comment by varshaprasad96** [comment]
@camilamacedo86  To include test for conversion WH in e2e, we may still need https://github.com/kubernetes-sigs/kubebuilder/pull/1644, because with v1beta1 crds, we still get an error: Invalid value: "Webhook": must be None if spec.preserveUnknownFields is true
The workaround would be to modify and set this separately, because including this in CRD_OPTIONS does not work either. Shall I implement the workaround in this test, or wait for the other PR to get merged and add test for this feature in the follow up (since in this PR we would only be adding the scaffolding) ?

**Comment by camilamacedo86** [comment]
I think is better to wait for #1644. Then, we can ensure that it works. Otherwise, will not the users face the same?

File: test/e2e/utils/test_context.go
Comment (camilamacedo86) [inline test/e2e/utils/test_context.go:71]:
```suggestion
		Domain:        fmt.Sprintf("example.com%s" + testSuffix),
```

File: test/e2e/utils/test_context.go
Comment (camilamacedo86) [inline test/e2e/utils/test_context.go:75]:
```suggestion
		Kind:           fmt.Sprintf("Foo%s" + testSuffix), 
```

File: test/e2e/utils/test_context.go
Comment (camilamacedo86) [inline test/e2e/utils/test_context.go:78]: Could we use the fmt.Sprintf to concat the values above?

File: testdata/project-v2/PROJECT
Comment (camilamacedo86) [inline testdata/project-v2/PROJECT:11]: Why we are changing v2 samples if the change is only for v3+?

File: test/e2e/utils/test_context.go
Comment (camilamacedo86) [inline test/e2e/utils/test_context.go:42]: IMO we sould not add Secondversion here. What does it means for the TestContext?

File: generate_testdata.sh
Comment (camilamacedo86) [inline generate_testdata.sh:69]: I did not find the scaffold in the project-v3 with the --hub/spoke content. Should not it change the scaffolds?

File: pkg/plugin/v3/webhook.go
Comment (varshaprasad96) [reply pkg/plugin/v3/webhook.go:84]: Since it was suggested that we preserve the current implementation, have modified this such that we can either specify `hub` and `spoke`(need not add `--conversion`) or just add `--conversion`. If former format is followed, we would scaffold implementations for hub & spoke, if not then do the same as it was followed previously.

File: test/e2e/utils/test_context.go
Comment (varshaprasad96) [reply test/e2e/utils/test_context.go]: Done.

File: test/e2e/utils/test_context.go
Comment (varshaprasad96) [reply test/e2e/utils/test_context.go]: Sorry, this was an intermediate change while testing. Have reverted it back.

File: generate_testdata.sh
Comment (varshaprasad96) [reply generate_testdata.sh]: Have modified the script. The scaffold appears now.

File: generate_testdata.sh
Comment (camilamacedo86) [inline generate_testdata.sh:62]: The commits needs to be squashed

File: pkg/plugin/v3/scaffolds/internal/templates/config/api/webhook.go
Comment (Adirio) [inline pkg/plugin/v3/scaffolds/internal/templates/config/api/webhook.go:44]: This is not used, right? Maybe some leftovers from previous commits.
```suggestion
```

File: pkg/plugin/v3/scaffolds/webhook.go
Comment (Adirio) [inline pkg/plugin/v3/scaffolds/webhook.go:107]:
```suggestion
	if err := machinery.NewScaffold().Execute(
		s.newUniverse(),
		&api.Webhook{Defaulting: s.defaulting, Validating: s.validation},
		&templates.MainUpdater{WireWebhook: true},
		&kdefault.InjectCAPatch{},
	); err != nil {
		return err
	}
	if s.conversion {
		if s.webhookSpokeVersion != "" && s.webhookHubVersion != "" {
			if err := machinery.NewScaffold().Execute(
				&api.HubConversion{HubVersion: s.webhookHubVersion},
				&api.SpokeConversion{SpokeVersion: s.webhookSpokeVersion},
			); err != nil {
				return err
			}
		} else {
			fmt.Println(`Webhook server has been set up for you.		
 You need to implement the conversion.Hub and conversion.Convertible interfaces for your CRD types.`)
			fmt.Println("Alternatively, Hub and Spoke version can be specified in the command to scaffold Convertible and Hub interfaces")
		}
	}
```

File: pkg/plugin/v3/webhook.go
Comment (Adirio) [inline pkg/plugin/v3/webhook.go:115]: Not worth a new dependency for this. Also, better if we move the message out of `Validate` (already did in the previous suggestion).
```suggestion
```

File: pkg/plugin/v3/webhook.go
Comment (Adirio) [inline pkg/plugin/v3/webhook.go:127]: Lets use `fmt.Errorf` not to import another library.
```suggestion
			return fmt.Errorf("Both Hub and Spoke version need to be specified when using conversion webhooks")
```

File: test/e2e/utils/test_context.go
Comment (Adirio) [inline test/e2e/utils/test_context.go:75]: Can you revert this change? It has nothing to do with this PR and is only done partially.

File: go.mod
Comment (Adirio) [inline go.mod:13]:
```suggestion
```

File: pkg/plugin/v3/webhook.go
Comment (Adirio) [inline pkg/plugin/v3/webhook.go:20]:
```suggestion
```

**Review by Adirio — CHANGES_REQUESTED** [review]
This PR needs as some of the packages have changed location. `pkg/plugin/v3` can now be found at `pkg/plugins/golang/v3`.

Also, I added several comments inline.

**Comment by camilamacedo86** [comment]
Hi @varshaprasad96, 

Is possible to release it with the master and see if we can get merged or is no longer it required?  WDYT?

**Comment by varshaprasad96** [comment]
@camilamacedo86 Ill resolve the conflicts and update with the most recent changes

**Comment by varshaprasad96** [comment]
Closing this PR in favor of #2084. This PR required some significant amount of changes based on v3-alpha plugin. Hence, started fresh and created a new PR for this feature on v3 plugin.

**Pull request closed without merging**
